presto-main/src/main/java/com/facebook/presto/operator/aggregation/AggregationCompiler.java

```diff
@@ -254,4 +232,209 @@ private static Method getCombineFunction(Class<?> clazz, Class<?> stateClass)
+    {
+        private final Signature signature;
+        private final String description;
+        private final boolean approximate;
+        private final boolean decomposable;
+
+        private final Class<?> definitionClass;
+        private final Class<?> stateClass;
+        private final Method inputFunction;
+        private final Method outputFunction;
+
+        public BindableAggregationFunction(Signature signature,
+                                           String description,
+                                           boolean approximate,
+                                           boolean decomposable,
+                                           Class<?> definitionClass,
+                                           Class<?> stateClass,
+                                           Method inputFunction,
+                                           Method outputFunction)
+        {
+            this.signature = signature;
+            this.description = description;
+            this.approximate = approximate;
+            this.decomposable = decomposable;
+            this.definitionClass = definitionClass;
+            this.stateClass = stateClass;
+            this.inputFunction = inputFunction;
+            this.outputFunction = outputFunction;
+        }
+
+        public Signature getSignature()
+        {
+            return signature;
+        }
+
+        public String getDescription()
+        {
+            return description;
+        }
+
+        public InternalAggregationFunction specialize(BoundVariables variables,
+                                                      TypeManager typeManager,
+                                                      int arity)
+        {
+            // bind variables
+            Signature boundSignature = SignatureBinder.bindVariables(signature, variables, arity);
+            List<Type> inputTypes = boundSignature.getArgumentTypes().stream().map(x -> typeManager.getType(x)).collect(toImmutableList());
+            Type outputType = typeManager.getType(boundSignature.getReturnType());
+
+            AggregationFunction aggregationAnnotation = definitionClass.getAnnotation(AggregationFunction.class);
+            requireNonNull(aggregationAnnotation, "aggregationAnnotation is null");
+
+            DynamicClassLoader classLoader = new DynamicClassLoader(definitionClass.getClassLoader());
+
+            AggregationMetadata metadata;
+            AccumulatorStateSerializer<?> stateSerializer = new StateCompiler().generateStateSerializer(stateClass, classLoader);
+            Type intermediateType = stateSerializer.getSerializedType();
+            Method intermediateInputFunction = getIntermediateInputFunction(definitionClass, stateClass);
+            Method combineFunction = getCombineFunction(definitionClass, stateClass);
+            AccumulatorStateFactory<?> stateFactory = new StateCompiler().generateStateFactory(stateClass, classLoader);
+
+            try {
```

Comment:
Lets move all the unreflection into the constructor. Otherwise, if you get an exception it won't show up until someone submits a query that tries to use the aggregation function. You might need some new structure to hold the partial parameter metadatas, since the input types aren't known in the constructor